Sources/ScipioKit/Producer/FrameworkProducer.swift

```diff
@@ -170,45 +167,55 @@ struct FrameworkProducer {
         return restored
     }
 
-    // Return true if pre-built artifact is available (already existing or restored from cache)
-    private func restore(target: CacheSystem.CacheTarget, cacheSystem: CacheSystem) async throws -> Bool {
-        let product = target.buildProduct
-        let frameworkName = product.frameworkName
-        let outputPath = outputDir.appendingPathComponent(product.frameworkName)
-        let exists = fileSystem.exists(outputPath.absolutePath)
-
-        if isConsumingCacheEnabled {
-            let expectedCacheKey = try await cacheSystem.calculateCacheKey(of: target)
-            let isValidCache = await cacheSystem.existsValidCache(cacheKey: expectedCacheKey)
-            let expectedCacheKeyHash = try expectedCacheKey.calculateChecksum()
-            if isValidCache && exists {
-                logger.info(
-                    "✅ Valid \(product.target.name).xcframework (\(expectedCacheKeyHash)) is exists. Skip building.", metadata: .color(.green)
-                )
-                return true
-            } else {
-                if exists {
-                    logger.warning("⚠️ Existing \(frameworkName) is outdated.", metadata: .color(.yellow))
-                    logger.info("🗑️ Delete \(frameworkName)", metadata: .color(.red))
-                    try fileSystem.removeFileTree(outputPath.absolutePath)
-                }
-                let restoreResult = await cacheSystem.restoreCacheIfPossible(target: target)
-                switch restoreResult {
-                case .succeeded:
-                    logger.info("✅ Restore \(frameworkName) (\(expectedCacheKeyHash)) from cache storage.", metadata: .color(.green))
+    /// Sendable interface to provide restore caches
+    private struct Restorer: Sendable {
+        let outputDir: URL
+        let cacheMode: Runner.Options.CacheMode
+        let fileSystem: any FileSystem
+
+        // Return true if pre-built artifact is available (already existing or restored from cache)
+        func restore(
+            target: CacheSystem.CacheTarget,
+            cacheSystem: CacheSystem
+        ) async throws -> Bool {
+            let product = target.buildProduct
+            let frameworkName = product.frameworkName
+            let outputPath = outputDir.appendingPathComponent(product.frameworkName)
+            let exists = fileSystem.exists(outputPath.absolutePath)
+
+            if cacheMode.isConsumingCacheEnabled {
```

Comment:
This comment is out of scope of this PR, but an early return is better in this case.
```swift
if !cacheMode.isConsumingCacheEnabled {
    return false
}

let expectedCacheKey = try await cacheSystem.calculateCacheKey(of: target)
// ...
```

(Please address this comment in another PR.)